### DISCUSSION TITLE
Variance computation in linen.LayerNorm

### DISCUSSION THREAD

**Discussion by stefano-1981** [google/flax#1709] [Q&A] [answered]
Hi there,

I am porting a model from pytorch to JAX. In the process, I am writing some regression tests to make sure ported modules match the original input-outputs.

I had some struggles with linen.LayerNorm. The problem came from `_compute_stats()`, where the variance is calculated as `Var = E[x^2] - E[x]^2 instead of Var = E[(x - E[x])^2])`. Now this can create a few issues for tensors that are not centered. I had a random tensor as input with these values:

```python
t = np.array([18348.604, 18348.08 , 18347.994, 18348.48 , 18348.088,
             18348.525, 18348.59 , 18347.863, 18348.666, 18347.822,
             18347.947, 18348.174, 18348.297, 18347.816, 18348.012,
             18348.447, 18348.72 , 18348.14 , 18348.387, 18348.496,
             18347.957, 18348.473, 18348.408, 18348.572, 18348.389,
             18348.342, 18348.363, 18347.94 , 18347.781, 18348.729,
             18348.523, 18348.207, 18347.797, 18348.637, 18348.242,
             18348.336, 18348.324, 18347.938, 18348.424, 18348.709,
             18347.883, 18347.836, 18348.72 , 18348.682, 18347.795,
             18348.096, 18348.549, 18348.754, 18348.479, 18348.271,
             18348.18 , 18348.686, 18347.844, 18348.441, 18348.61 ,
             18348.508, 18348.24 , 18347.94 , 18347.781, 18348.059,
             18348.287, 18347.963, 18348.281, 18347.86 , 18347.781,
             18348.041, 18348.232, 18348.559, 18348.213, 18347.799,
             18348.59 , 18347.818, 18348.213, 18348.703, 18348.127,
             18348.701, 18348.441, 18348.297, 18348.318, 18348.443,
             18348.482, 18348.572, 18347.94 , 18347.9  , 18347.965,
             18347.834, 18348.057, 18348.238, 18347.867, 18348.28 ,
             18347.896, 18348.094, 18348.477, 18348.3  , 18347.975,
             18347.885, 18348.488, 18348.   , 18348.373, 18348.414,
             18348.758, 18348.72 , 18348.13 , 18348.742, 18348.7  ,
             18348.004, 18348.248, 18347.816, 18348.44 , 18348.434,
             18348.188, 18348.4  , 18348.484, 18348.   , 18348.244,
             18347.8  , 18348.213, 18347.877, 18348.215, 18348.22 ,
             18347.809, 18348.596, 18347.95 , 18348.367, 18348.656,
             18348.012, 18347.895, 18348.678], np.float32)
```

and the variance calculation for np.float32 is quite off:

```python
print(np.var(t))  # -> 0.088308185
print(np.mean(np.square(t), -1) - np.square(np.mean(t, -1))) # -> -32.0
```
In this particular case the value is negative so it gets thresholded at zero in `_compute_stats()`, but I got also +32, in other cases. Changing the precision to np.float64 would solves the problem, but the numerical problem seems also solved when the tensor is centered (without increasing precision to float64):

```python
z = t - np.mean(t)
print(np.var(z))  # -> 0.088307485
print(np.mean(np.square(z), -1) - np.square(np.mean(z, -1)))  # -> 0.08830748
```
Would it make sense to add such a centering in the `_compute_stats()`? Is there anything else I should be doing, instead?

**Comment by marcvanzee** [comment, marked as answer]
`E[x^2] - E[x]^2` is an approximation of `E[(x - E[x])^2])`, so it is expected that you will get some difference. The reason why we are using the approximation is that we are currently trading efficiency for precision.  We can compute the two clauses of `E[x^2] - E[x]^2` in parallel in a single XLA fusion, which is faster than first computing `a = E[x]` and then `E[(x - a)^2])` because this cannot be parallelized.

Given how floating points are represented in memory, this difference will increase as the exponent of the floating point numbers increases. Given that 1) Your exponents are quite large and the differences between the numbers relatively small, and 2) you are computing in half precision, this difference will be quite large.

I do feel there is something to be said for allowing the other option as well, since I can imagine users might prefer precision over efficiency (as you seem to do). @jheek WDYT of adding this as an option?
